plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -431,28 +432,42 @@ spendBudgetCek = let (CekBudgetSpender spend) = ?cekBudgetSpender in spend
 -- see Note [Scoping].
 -- | Instantiate all the free variables of a term by looking them up in an environment.
 -- Mutually recursive with dischargeCekVal.
-dischargeCekValEnv :: CekValEnv uni fun -> Term Name uni fun () -> Term Name uni fun ()
-dischargeCekValEnv !valEnv =
-    -- We recursively discharge the environments of Cek values, but we will gradually end up doing
-    -- this to terms which have no free variables remaining, at which point we won't call this
-    -- substitution function any more and so we will terminate.
-    termSubstFreeNames $ \name -> do
-        val <- lookupName name valEnv
-        Just $ dischargeCekValue val
+dischargeCekValEnv :: forall uni fun. CekValEnv uni fun -> Term NamedDeBruijn uni fun () -> Term NamedDeBruijn uni fun ()
+dischargeCekValEnv valEnv = go 0
+ where
+  -- The lamCnt is just a counter that measures how many lambda-abstractions
+  -- we have descended in the `go` loop.
+  go :: Word -> Term NamedDeBruijn uni fun () -> Term NamedDeBruijn uni fun ()
+  go !lamCnt =  \case
+    LamAbs ann name body -> LamAbs ann name $ go (lamCnt+1) body
+    var@(Var _ (NamedDeBruijn _ ndbnIx)) -> let ix = coerce ndbnIx :: Word  in
+        if lamCnt >= ix
+        -- the index n is less-than-or-equal than the number of lambdas we have descended
+        -- this means that n points to a bound variable, so we don't discharge it.
+        then var
+        else
+            -- index relative to (as seen from the point of view of) the environment
+            dischargeCekValue $ Env.unsafeIndex valEnv $ ix - lamCnt
```

Comment:
We should definitely not use `unsafeIndex`!

Reply:
I know, and I wanted to discuss about this